### ISSUE TITLE
Develop input system for plot data

### ISSUE THREAD

**Issue by sacrevert** [BiologicalRecordsCentre/BSBI-Card-and-PlantPortal-DEPRECATED-#4]
Latest versions of spec and work plan here:
[Plant Portal Work Plan_v2016 06 09.docx](https://github.com/BiologicalRecordsCentre/PlantPortal/files/306645/Plant.Portal.Work.Plan_v2016.06.09.docx)
[PlantPortalDevelopment_v2016 05 31.docx](https://github.com/BiologicalRecordsCentre/PlantPortal/files/306650/PlantPortalDevelopment_v2016.05.31.docx)

Nested in #46

**Comment by andrewvanbreda** [comment]
I took a look at your new comments in the work plan. There isn't anything in there that is unexpected. You are correct that a user bulk import would want to use the same code as a large historical import, as essentially to the computer they are the same (or at least very similar) things. This makes life much easier, in particular testing effort and bug fix effort are reduced as essentially fixing a bug or testing results in that activity being applied to both. It might also be possible to have a situation where one might be in a web interface and one might be running the SQL directly, but again even in this situation they could share code.

**Comment by sacrevert** [comment]
Latest version of spec, phone meeting 07 July 2016 and
[Plant Portal Skype Meeting Notes – 7th July 2016.docx](https://github.com/BiologicalRecordsCentre/PlantPortal/files/356722/Plant.Portal.Skype.Meeting.Notes.7th.July.2016.docx)
[AvB comments on Skype issues 11th July 2016.docx](https://github.com/BiologicalRecordsCentre/PlantPortal/files/356723/AvB.comments.on.Skype.issues.11th.July.2016.docx)

[PlantPortalDevelopment_v2016 06 17.docx](https://github.com/BiologicalRecordsCentre/PlantPortal/files/356719/PlantPortalDevelopment_v2016.06.17.docx)

**Comment by andrewvanbreda** [comment]
I have a question regarding the importer. I have a pretty good idea of how the grouping of the samples/plots can work. However having started coding the importer I have noticed an issue, how do we decide if the sample or plot being imported should be created as a new one, or try to use an existing one in the database. We could just try to only use the existing plot locations if all the fields match including Plot Name and Grid Ref and Group ID, ask the user for confirmation if existing location detection? 
I suppose we could do something similar with samples.
This would probably coming in the form of a grid screen displaying all the records where existing matches are found and asking uses to confirm use existing with checkboxes down the side.
Am not currently quite sure how to deal with this currently as this starts straying from the original importer this is being based on which works heavily with the warehouse, which is more difficult to change.

The alternative would be to always assume the samples are new, but try to do the match on the existing plots only.

The other problem is in the import file itself, how do we decide when to create new samples and plots. 
For instance we might have 10 occurrence rows including matching sample information for each row. How will we know if those 10 really are for the same sample.
I think before we spoke about using a sample id on the rows so we know when to create a new sample. 
This would be different from the group id/name, is the above solution OK to go forward with, or is it too confusing (as confusing as this post perhaps!)

**Comment by sacrevert** [comment]
Hi Andy, when you say "how do we decide if the sample or plot being imported should be created as a new one, or try to use an existing one in the database" I presume you are talking about sample/plot groupings only. If so, I think that the method of testing to see whether that combination of plot name (or ID), grid ref (or other spatial ref) and plot group ID already exist, and the same for samples (i.e. testing whether sample ID, grid/spatial ref and sample group ID already exist. I presume that that the on-screen confirmation would then be a matter of selecting which plot grouping and/or sample grouping was intended (if any).

I note here that, for some historic data, it might be the case that there is considerable overlap between datasets. For example, if the spatial ref is ‘Britain’, and a single plot group ID is given (e.g. ‘1’), and plot name/numbering starts from 1. I don’t think this is that likely, but just flag it as a possibility. I suppose this is not a problem, but I just note that there will be a need to assign truly unique plot group, plot, and sample IDs to data. I imagine that as well as assigning unique IDs, we would also want to display the ID numbers assigned by the user (although we could always remind the user that there IDs are not unique, and suggest changing them for their own benefit (but this would not be essential for the database)).

Regarding the second question, is this not solved by asking the user to specify unique sample IDs? So, to clarify, we would be asking the user to specify (i) sample ID (i.e. occurrence groupings); (ii) sample group IDs (which might indicate samples that they wanted to store together, e.g. samples in a particular field or habitat patch, presumably made at a similar time); (iii) plot group IDs (I suppose this could be project metadata, i.e. only one group allowed, although people might want to group physical plot locations within a project, I'm not sure that there is any benefit from letting them do so within the context of our Plant Portal); and also (iv) plot IDs.

**Comment by andrewvanbreda** [comment]
Hi Oli,

OK, I thought we would just use the Sample ID during the import of rows and then not store it, but you are right, it is better if we store that too.
OK, I can see us getting confused with terminology here as there is also a Sample ID in the database sense, what we are talking about is the names the users will be entering to identify the samples.
So let's refer to these as
-Sample identifier name
-Sample group identifier name
-Plot group identifier name

Yes, so if we actually store a Sample identifier name against the sample we can use that to check whether to use it as an existing sample (and also make sure the rest of the sample data in the import is consistent).
Part of that check would obviously have to be on the sample plot anyway.
So someone could import 10 occurrence rows on an existing sample.

I think you have answered my second question with regards to the plots to check an exact match on the Plot Name and Grid Ref and Group ID, as Plots don't have their own unique id (other than the database one which isn't visible to the users)

So if someone imports some samples, we can still check to see if the new sample is using an existing plot by checking those fields.

Thanks

**Comment by DavidRoy** [comment]
Is there also a role for the ‘External Key’ field in the Indicia Samples table?

**Comment by sacrevert** [comment]
OK. Thanks Andy. Just to say that I think the idea of restricting the Plot group identifier name to a single value might be worthwhile, then this could be described in documentation as a project ID, and would make it clearer that new samples were to be added to a particular project (i.e. set of plots). I can't really see a lot of value in letting users specify multiple subgroups of plots within a project if they can already group samples. Plus downloads could allow for the selection of plots/samples within projects anyway, without need for actual subsets of plots specified in the database.

**Comment by andrewvanbreda** [comment]
In answer to David's query about the external_key.

I have not used the external key on the sample (or locations) table before (just on the termlist table)

I think normally we would use it as a unique id to identify the sample/location (although it isn't enforced to be unique in the database). 
I suppose it could store the Sample identifier name, or even the Sample/Plot group identifier names if we don't mind the external_key not being unique.

The only things that I can think of that might be drawbacks of doing it the external key way is if we are doing everything else using attributes, then it might be more consistent save the identifiers using attributes too.
Also is there a specified format that the sample or location external key would normally be. As I didn't think we are enforcing a format on the users' sample group identifier name/location group identifier name, or are we?
The other thing I can think of is if we use the external_key, then that field can no longer be used for anything else further down the line.

I don't know, I suppose if anyone can point out a definite advantage of using the external_key field then I am all ears.

In response to Oli's comment. That is fine, the Plot group identifier name can be single value. We can call it a project ID if you like, if that is the way plots are going to be grouped (on a project basis).
I was just trying to avoid us getting confused with the database meaning of ID.

**Comment by sacrevert** [comment]
I'm afraid I can't comment on the use or otherwise of the external_key field. However I think that all IDs (in the database sense) should be enforced to be unique (although I know that we have agreed that non-unique user-assigned identifiers will be stored and displayed as well). Unique database IDs might be important in the long run if the scientific community want to use them in data archiving related to scientific papers.

On the second point, Plot group identifier name is fine. let us keep using the terms you suggested, I just added project ID in an informal sense to aid understanding (!), I'll try to avoid using this term in the future.

**Comment by andrewvanbreda** [comment]
Yes, when I meant not unique, I was just referring to the external_key field, if for instance we used this field to store the group identifier then it would be the same for all the samples or plots in the group and therefore not unique for the sample or plot (location).
I think for now I will press ahead with storing the identifiers as attributes anyway, as at the moment I don't see any obvious disadvantage to doing it that way.

**Comment by DavidRoy** [comment]
Agreed, thanks

**Comment by andrewvanbreda** [comment]
I have spoken to John about my suggestions for the database and after much thought we have decided those initial suggestions I made are the best way to go apart from one change listed below, and also one question also listed below.
1. A slight alteration to the plan is that where we have attributes that need to be associated with the species and the community.
   My initial suggestions was to encode these attributes into the text_value field for a community/taxon attribute value row.
   This is how the current complex attributes work.
   However john was a bit concerned about performance of querying when we have to collect values that are encoded into a text field (using JSON).
   He has now suggested to save these attributes on separate rows in the database, but we will have a new column in the database displaying an ID number which will link the rows together so we know the attributes are grouped together.
2. The other question that has arisen is about the sample/plot groups. The basic idea I have of storing the groups as terms, then making use of the new term attributes to store information about the groups seems OK.
   However there is a slight concern about what these groupings are for?
   For instance, would you ever want to assign actually people to groups of samples? John seemed to think sample groups might be a bit like recording groups.
   Basically we need to make sure we can actual support any functionality you are going to want to associate with the groupings going forward into the future.

**Comment by sacrevert** [comment]
1. Ok, this sounds like a good idea.
2. Plot groups are basically for organising plots into projects, and so for the storage of project metadata, as described in other posts above. Sample groups (whether at one point in time or across years) are so that users can keep tabs on samples that were in a particular habitat patch; this might also become more directly useful in the future if we enable the user to send sample groups off to another web service for analysis (e.g. community classifications often require sets of 5 or 10 plots for analysis).
   Regarding the assignment of people, to groups, I suppose this would be useful if it was the only way of allowing another user to edit or analyse plots -- is that the case? I had envisaged that some users would get verifier status and have rights to edit occurrences, and we might want to allow people to analyse or download data that they did not create -- does this have any bearing on whether to allow the option of assigning people to plot or sample groups?

**Comment by andrewvanbreda** [comment]
I suppose the only issue I have here is that we might want to assign sample groups for a user at some point in the future so that user is able to perform some operation on the samples (editing or whatever). I suppose in that case we  could assign the group in the same way as a square is on NPMS (by saving the ID of the item to assign against the person in the database).
I will speak to John about this, has he was the one who queried it more.
